lib/heap/cmpctmalloc/cmpctmalloc.c

```diff
@@ -0,0 +1,608 @@
+		FixLeftPointer(right, (header_t *)free);
+		head->header.size -= left_over;
+	} else {
+		unlink_free(head, bucket);
+	}
+	void *result =
+		create_allocation_header(head, 0, head->header.size, head->header.left);
+#ifdef DEBUG
+	memset(result, ALLOC_FILL, size - sizeof(header_t));
+	memset(((char *)result) + size - sizeof(header_t), PADDING_FILL, rounded_up - size);
+#endif
+	unlock();
+	return result;
+}
+
+void *cmpct_memalign(size_t size, size_t alignment)
+{
+	if (alignment < 8) return cmpct_alloc(size);
+	size_t padded_size =
+		size + alignment + sizeof(struct free_heap_chunk) + sizeof(header_t);
+	char *unaligned = (char *)cmpct_alloc(padded_size);
+	lock();
+	size_t mask = alignment - 1;
+	uintptr_t payload_int = (uintptr_t)unaligned + sizeof(struct free_heap_chunk) +
+		sizeof(header_t) + mask;
+	char *payload = (char *)(payload_int & ~mask);
+	if (unaligned != payload) {
+		header_t *unaligned_header = (header_t *)unaligned - 1;
+		header_t *header = (header_t *)payload - 1;
+		size_t left_over = payload - unaligned;
+		create_allocation_header(
+			header, 0, unaligned_header->size - left_over, unaligned_header);
+		header_t *right = right_header(unaligned_header);
+		unaligned_header->size = left_over;
+		FixLeftPointer(right, header);
+		unlock();
+		cmpct_free(unaligned);
+	} else {
+		unlock();
+	}
+	// TODO: Free the part after the aligned allocation.
+	return payload;
+}
+
+void cmpct_free(void *payload)
+{
+	if (payload == NULL) return;
+	header_t *header = (header_t *)payload - 1;
+	size_t size = header->size;
+	lock();
+	header_t *left = header->left;
+	if (left != NULL && is_tagged_as_free(left)) {
+		// Coalesce with left free object.
+		unlink_free_unknown_bucket((struct free_heap_chunk *)left);
+		header_t *right = right_header(header);
+		if (is_tagged_as_free(right)) {
+			// Coalesce both sides.
+			unlink_free_unknown_bucket((struct free_heap_chunk *)right);
+			header_t *right_right = right_header(right);
+			FixLeftPointer(right_right, left);
+			free_memory(left, left->left, left->size + size + right->size);
+		} else {
+			// Coalesce only left.
+			FixLeftPointer(right, left);
+			free_memory(left, left->left, left->size + size);
+		}
+	} else {
+		header_t *right = right_header(header);
+		if (is_tagged_as_free(right)) {
+			// Coalesce only right.
+			header_t *right_right = right_header(right);
+			unlink_free_unknown_bucket((struct free_heap_chunk *)right);
+			FixLeftPointer(right_right, header);
+			free_memory(header, left, size + right->size);
+		} else {
+			free_memory(header, left, size);
+		}
+	}
+	unlock();
+}
+
+void *cmpct_realloc(void *payload, size_t size)
+{
+	if (payload == NULL) return cmpct_alloc(size);
+	header_t *header = (header_t *)payload - 1;
+	size_t old_size = header->size - sizeof(header_t);
+	void *new_payload = cmpct_alloc(size);
+	memcpy(new_payload, payload, MIN(size, old_size));
+	cmpct_free(payload);
+	return new_payload;
+}
+
+void cmpct_delayed_free(void *ptr)
```

Comment:
The delayed free stuff is all up in lib/heap/heap_wrapper.c now, so all of that can be removed out of this file. Far as I can tell no one calls this anyway.